preview/MsixCore/msixmgr/Package.cpp

```diff
@@ -154,20 +160,64 @@ HRESULT MsixCoreLib::PackageBase::ProcessPSFIfNecessary()
-
-            // resolve the given paths from json into full paths.
-            m_executionInfo.resolvedExecutableFilePath = FilePathMappings::GetInstance().GetExecutablePath(psfExecutable, m_packageFullName.c_str());
-            m_executionInfo.workingDirectory = FilePathMappings::GetInstance().GetExecutablePath(psfWorkingDirectory, m_packageFullName.c_str());
-
-            TraceLoggingWrite(g_MsixTraceLoggingProvider,
-                "PSF redirection",
-                TraceLoggingWideString(m_executionInfo.resolvedExecutableFilePath.c_str(), "Resolved PSF executable"),
-                TraceLoggingWideString(m_executionInfo.workingDirectory.c_str(), "WorkingDirectory"),
-                TraceLoggingWideString(m_executionInfo.commandLineArguments.c_str(), "Arguments"));
+            std::wstring jsonFile = p.path();
+            std::ifstream ifs(jsonFile);
+            IStreamWrapper isw(ifs);
+
+            Document doc;
+            doc.ParseStream(isw);
+
+            if (doc.HasParseError())
+            {
+                //return error
+            }
+
+            const Value& apps = doc["applications"];
+
+            for (Value::ConstValueIterator itr = apps.Begin(); itr != apps.End(); ++itr)
+            {
+                std::string id = (*itr)["id"].GetString();
+                std::wstring jsonApplicationId;
+                jsonApplicationId.assign(id.begin(), id.end());
+
+                //ApplicationId in json should match applicationId from the xml manifest. Right now, we only process first application from the xml manifest anyway
+                if (CaseInsensitiveEquals(jsonApplicationId, m_applicationId))
+                {
+                    std::string executable = (*itr)["executable"].GetString();
+                    std::wstring psfExecutable;
+                    psfExecutable.assign(executable.begin(), executable.end());
+
+                    std::string arguments = (*itr)["arguments"].GetString();
+                    std::wstring psfArguements;
+                    psfArguements.assign(arguments.begin(), arguments.end());
+
+                    std::string workingDirectory = (*itr)["workingDirectory"].GetString();
+                    std::wstring psfWorkingDirectory;
+                    psfWorkingDirectory.assign(workingDirectory.begin(), workingDirectory.end());
+
+                    m_executionInfo.commandLineArguments = psfArguements;
+
+                    // resolve the given paths from json into full paths.
+                    m_executionInfo.resolvedExecutableFilePath = FilePathMappings::GetInstance().GetExecutablePath(psfExecutable, m_packageFullName.c_str());
+                    m_executionInfo.workingDirectory = FilePathMappings::GetInstance().GetExecutablePath(psfWorkingDirectory, m_packageFullName.c_str());
+
+                    const Value& startScript = (*itr)["startScript"];

```

Comment:
>startScript [](start = 55, length = 11)

we also want to parse runOnce value, and endScript. In the case of runOnce=false or if endScript is present, we should fail the install, as we do not currently support those cases.

Reply:
If there is an error in parsing the json, should the install be failed in that case as well?

Reply:
yeah. wouldn't that be caught earlier in doc.HasParseError()? you have a comment there to return an error. also keep in mind that it's ok for startScript to be missing entirely from the json. 

---
In reply to: [340250145](https://github.com/microsoft/msix-packaging/pull/249#discussion_r340250145) [](ancestors = 340250145)

Reply:
we should probably also error out if the appplicationId cannot be found

---
In reply to: [340253280](https://github.com/microsoft/msix-packaging/pull/249#discussion_r340253280) [](ancestors = 340253280,340250145)

Reply:
Added checks for runOnce, endScript, return if error in parsing json, and erroring out(return ELEMENT_NOT_FOUND) if app is not found in the applications array at all.